### PR TITLE
[teleport-update] needrestart and systemd drop-in

### PR DESCRIPTION
**Pull request by sclevine**
This PR sets the default needrestart configuration to disabled for the Teleport systemd service. This PR only affects Teleport when it is installed via the `teleport-update` command, and does not change the behavior of the teleport deb/rpm package.

needrestart is a Debian service that restarts services when their underlying libraries change. Networking services like Teleport are not always safe to restart during package upgrades, because they may disrupt connections. If needed, users may still opt-in to enable restarts for Teleport via dpkg (including interactive prompting). [Internal link with more details](https://gravitational.slack.com/archives/C04P78M46F2/p1733334395335229).

This PR also adds a systemd drop-in that `teleport-update` can use to provide env vars to `teleport`. This will let `teleport` read agent group identifiers in the future. It's included here because the logic is very similar to the needrestart feature, and both features drove out some minor refactoring.

```
ubuntu@legendary-mite:~$ cat /etc/systemd/system/teleport.service.d/teleport-update.conf 
# teleport-update
# DO NOT EDIT THIS FILE
[Service]
Environment=TELEPORT_UPDATE_CONFIG_FILE=/opt/teleport/default/update.yaml

ubuntu@legendary-mite:~$ cat /etc/needrestart/conf.d/teleport-update.conf 
$nrconf{override_rc}{qr(^teleport\.service)} = 0;
```

---

The `teleport-update` binary will be used to enable, disable, and trigger automatic Teleport agent updates. The new auto-updates system manages a local installation of the cluster-specified version of Teleport stored in `/opt/teleport`.

RFD: https://github.com/gravitational/teleport/pull/47126
Goal (internal): https://github.com/gravitational/cloud/issues/10289

### CODE REVIEW

**Comment by sclevine** [comment]
Reviews appreciated when anyone has an opportunity 🙂

File: lib/autoupdate/agent/setup.go [hunk 1]
```diff
@@ -204,11 +222,15 @@ func (ns *Namespace) Teardown(ctx context.Context) error {
...
-	if err := os.Remove(ns.updaterServiceFile); err != nil && !errors.Is(err, fs.ErrNotExist) {
-		return trace.Wrap(err, "failed to remove teleport-update systemd service")
-	}
-	if err := os.Remove(ns.updaterTimerFile); err != nil && !errors.Is(err, fs.ErrNotExist) {
-		return trace.Wrap(err, "failed to remove teleport-update systemd timer")
+	for _, p := range []string{
+		ns.updaterServiceFile,
+		ns.updaterTimerFile,
+		ns.dropInFile,
+		ns.needrestartConfFile,
```
Comment (vapopov) [inline lib/autoupdate/agent/setup.go:229]: should we ignore removal this config if it is not exist? because when we write template we ignore write if there are no directory

File: lib/autoupdate/agent/setup.go [hunk 2]
```diff
@@ -258,7 +316,11 @@ func writeTemplate(path, t string, values any) error {
 	}
 	defer f.Cleanup()
 
-	tmpl, err := template.New(file).Parse(t)
+	tmpl, err := template.New(file).Funcs(template.FuncMap{
+		"replace": func(s, old, new string) string {
+			return strings.Replace(s, old, new, -1)
```
Comment (vapopov) [inline lib/autoupdate/agent/setup.go:321]: nit: `strings.ReplaceAll` little bit simpler to read, so you don't need to go to read godoc

**Review by vapopov — APPROVED** [review]

**Comment by sclevine** [comment]
@codingllama @greedy52 @hugoShaka looking for another review when you have a chance 🙂

**Review by hugoShaka — APPROVED** [review]
We had an interesting discussion about what to do with needrestart and ended up with this solution. Could you summarize it in a comment somewhere in `setup.go` so we have a written trace about this decision and why we chose to opt-out instead of teaching needrestat to do soft reloads?

File: lib/autoupdate/agent/setup.go [hunk 3]
```diff
@@ -219,27 +241,63 @@ func (ns *Namespace) Teardown(ctx context.Context) error {
...
-	err = writeTemplate(
-		ns.updaterTimerFile, updateTimerTemplate,
-		struct{ TeleportService string }{filepath.Base(ns.serviceFile)},
+	teleportService := filepath.Base(ns.serviceFile)
+	err = writeTemplate(ns.updaterTimerFile, updateTimerTemplate,
+		struct {
+			TeleportService string
+		}{
+			teleportService,
+		},
```
Comment (hugoShaka) [inline lib/autoupdate/agent/setup.go:265]: optional nit: this feels a bit awkward for the caller to create anonymous strucs, would it be more readable if we had the structs defined next to their templates? This way, when one edits the template, they also update the struct containing the template input.

**Comment by sclevine** [comment]
> summarize it in a comment somewhere in setup.go

👍 Great idea, will do.
